File: packages/harness/src/lexical/script-utils.ts
/**
 * Script-aware text utilities shared by every lexical (non-LLM) check in this monorepo — the one
 * place CJK-vs-Latin handling lives, so a check's own regex/keyword logic doesn't need its own
 * tokenization strategy. Mirrored in adapter/harness/script_utils.py for Python's own harness
 * implementation; packages/personal-assistant imports this module (via @buildaharness/harness,
 * already a real dependency) rather than reimplementing it — personal-assistant is one
 * application built on buildaharness, not a place for harness-core infrastructure to live.
 *
 * `.split(/\s+/)`-based word counting/tokenization silently breaks on CJK text, which has no
 * inter-word whitespace — an entire Chinese sentence splits into exactly one "word". These
 * utilities give every consumer a tokenization that's meaningful for both: CJK spans are split
 * per-character (a cheap, no-dependency approximation of real segmentation — weaker precision
 * than a real segmenter, but every consumer already treats its own lexical check as a fast
 * pre-filter with the real judgment call handled elsewhere), non-CJK spans keep ordinary
 * whitespace-based word splitting.
 */

// CJK Unified Ideographs, Extension A, and Compatibility Ideographs — covers the vast majority of
// real-world Chinese/Japanese-kanji text without pulling in a full Unicode script database.
const CJK_CHAR = /[㐀-䶿一-鿿豈-﫿]/
// Unambiguous CJK clause/sentence boundaries (period, exclamation, question mark, semicolon —
// both fullwidth and halfwidth forms). Deliberately excludes the CJK comma (，) and enumeration
// comma (、): those are closer to an English comma and need the same conjunction-word
// disambiguation English commas already get elsewhere, not a blanket split.
const CJK_CLAUSE_PUNCTUATION = /[。！？；!?;]/g

export function containsCJK(text: string): boolean {
  return CJK_CHAR.test(text)
}

/**
 * Splits `text` into tokens: each CJK character becomes its own token, everything else is
 * whitespace-split as usual. For text with no CJK characters at all, this is byte-for-byte
 * equivalent to `text.split(/\s+/).filter(Boolean)` — the existing behavior every consumer being
 * migrated onto this module already relies on.
 */
export function tokenize(text: string): string[] {
  const tokens: string[] = []
  let buffer = ''
  for (const ch of text) {
    if (CJK_CHAR.test(ch)) {
      if (buffer) {
        tokens.push(...buffer.split(/\s+/).filter(Boolean))
        buffer = ''
      }
      tokens.push(ch)
    } else {
      buffer += ch
    }
  }
  if (buffer) tokens.push(...buffer.split(/\s+/).filter(Boolean))
  return tokens
}

/** Token count — keeps `WORD_LIMIT`-style length thresholds meaningful once CJK content exists. */
export function tokenCount(text: string): number {
  return tokenize(text).length
}

/**
 * Case-insensitive shared-token overlap between two strings, minus `stopwords` — the "do these
 * two statements share a subject" gate several lexical checks use before treating any other
 * signal (a negation pair, a trigger word) as meaningful. Filtering stopwords from each side
 * before intersecting (rather than filtering the intersection afterward) is mathematically
 * equivalent, since a stopword excluded from either side can never survive an intersection anyway.
 */
export function sharedTokens(a: string, b: string, stopwords: ReadonlySet<string> = new Set()): string[] {
  const setA = new Set(tokenize(a).map((t) => t.toLowerCase()).filter((t) => !stopwords.has(t)))
  const setB = new Set(tokenize(b).map((t) => t.toLowerCase()).filter((t) => !stopwords.has(t)))
  return [...setA].filter((t) => setB.has(t))
}

/**
 * Splits `text` into clauses on CJK sentence-ending punctuation and, additionally, on
 * `extraBoundary` if supplied — a caller's own English conjunction-word/punctuation boundary
 * (e.g. fact-extraction.ts's `CLAUSE_BOUNDARY`, risk-classifier.ts's `RISK_CLAUSE_BOUNDARY`),
 * applied within each CJK-punctuation-delimited chunk. For text with no CJK punctuation, this
 * reduces to exactly `text.split(extraBoundary)`.
 */
export function splitClauses(text: string, extraBoundary?: RegExp): string[] {
  const chunks = text.split(CJK_CLAUSE_PUNCTUATION)
  const result: string[] = []
  for (const chunk of chunks) {
    if (extraBoundary) result.push(...chunk.split(extraBoundary))
    else result.push(chunk)
  }
  return result.map((s) => s.trim()).filter(Boolean)
}

File: adapter/harness/script_utils.py
"""
Script-aware text utilities shared by every lexical (non-LLM) check in the Python harness
implementation — mirrors packages/harness/src/lexical/script-utils.ts function-for-function.

`.split()`-based word counting/tokenization silently breaks on CJK text, which has no
inter-word whitespace — an entire Chinese sentence splits into exactly one "word". These
utilities give every consumer a tokenization that's meaningful for both: CJK spans are split
per-character (a cheap, no-dependency approximation of real segmentation), non-CJK spans keep
ordinary whitespace-based word splitting.
"""

from __future__ import annotations

import re

# CJK Unified Ideographs, Extension A, and Compatibility Ideographs — matches
# script-utils.ts's CJK_CHAR range.
_CJK_CHAR = re.compile(r"[㐀-䶿一-鿿豈-﫿]")
# Unambiguous CJK clause/sentence boundaries (period, exclamation, question mark, semicolon —
# both fullwidth and halfwidth forms). Deliberately excludes the CJK comma (，) and enumeration
# comma (、) — see script-utils.ts's own comment for why.
_CJK_CLAUSE_PUNCTUATION = re.compile(r"[。！？；!?;]")


def contains_cjk(text: str) -> bool:
    return bool(_CJK_CHAR.search(text))


def tokenize(text: str) -> list[str]:
    """Splits `text` into tokens: each CJK character becomes its own token, everything else is
    whitespace-split as usual. For text with no CJK characters at all, this is identical to
    `text.split()` — the existing behavior every consumer being migrated onto this module
    already relies on.
    """
    tokens: list[str] = []
    buffer = ""
    for ch in text:
        if _CJK_CHAR.match(ch):
            if buffer:
                tokens.extend(buffer.split())
                buffer = ""
            tokens.append(ch)
        else:
            buffer += ch
    if buffer:
        tokens.extend(buffer.split())
    return tokens


def token_count(text: str) -> int:
    return len(tokenize(text))


def shared_tokens(a: str, b: str, stopwords: frozenset[str] = frozenset()) -> list[str]:
    """Case-insensitive shared-token overlap between two strings, minus `stopwords`."""
    set_a = {t.lower() for t in tokenize(a) if t.lower() not in stopwords}
    set_b = {t.lower() for t in tokenize(b) if t.lower() not in stopwords}
    return list(set_a & set_b)


def split_clauses(text: str, extra_boundary: re.Pattern[str] | None = None) -> list[str]:
    """Splits `text` into clauses on CJK sentence-ending punctuation and, additionally, on
    `extra_boundary` if supplied, applied within each CJK-punctuation-delimited chunk.
    """
    chunks = _CJK_CLAUSE_PUNCTUATION.split(text)
    result: list[str] = []
    for chunk in chunks:
        if extra_boundary is not None:
            result.extend(extra_boundary.split(chunk))
        else:
            result.append(chunk)
    return [s.strip() for s in result if s.strip()]
